File: 2024/Day02.py
from puzzle_input import puzzle_input


def check_report(report: list[int]) -> bool:
    diffs = [a - b for a, b in zip(report, report[1:])]
    return (all(d < 0 for d in diffs) or all(d > 0 for d in diffs)) and all(
        1 <= abs(d) <= 3 for d in diffs
    )


def check_report_pops(report: list[int]) -> bool:
    if check_report(report):
        return True
    for i in range(len(report)):
        r = report.copy()
        r.pop(i)
        if check_report(r):
            return True
    return False


if __name__ == "__main__":
    PUZZLE_INPUT = puzzle_input().splitlines()

    REPORTS = [[int(i) for i in line.split()] for line in PUZZLE_INPUT]

    print("Part 1:", sum(map(check_report, REPORTS)))
    print("Part 2:", sum(map(check_report_pops, REPORTS)))

File: 2024/Day02.hs
import PuzzleInput

parseLine :: (Num a, Read a) => String -> [a]
parseLine = map read . words

diffs :: (Num a) => [a] -> [a]
diffs rs = zipWith (-) rs (tail rs)

checkReport :: (Num a, Ord a) => [a] -> Bool
checkReport rs = (all (< 0) d || all (> 0) d) && all ((1 <=) . abs) d && all ((<= 3) . abs) d
  where
    d = diffs rs

getReportPop :: (Num a, Ord a) => [a] -> [[a]]
getReportPop (r : rs) = rs : map (r :) (getReportPop rs)
getReportPop [] = []

checkReportPop :: (Num a, Ord a) => [a] -> Bool
checkReportPop rs = checkReport rs || any checkReport (getReportPop rs)

countTrue :: [Bool] -> Int
countTrue = length . filter id

main :: IO ()
main = do
  contents <- puzzleInputLines
  let reports = map parseLine contents
  print (countTrue . map checkReport $ reports)
  print (countTrue . map checkReportPop $ reports)
